Docs/Proposals/0001-jsonencoder-key-strategy.md

```diff
@@ -0,0 +1,152 @@
+        case custom(([CodingKey]) -> CodingKey)
+    }
+    
+    /// The strategy to use for decoding keys. Defaults to `.useDefaultKeys`.
+    open var keyDecodingStrategy: KeyDecodingStrategy = .useDefaultKeys
+}
+
+class JSONEncoder {
+    /// The strategy to use for automatically changing the value of keys before encoding.
+    public enum KeyEncodingStrategy {
+        /// Use the keys specified by each type. This is the default strategy.
+        case useDefaultKeys
+        
+        /// Convert from "camelCaseKeys" to "snake_case_keys" before writing a key to JSON payload.
+        ///
+        /// Capital characters are determined by testing membership in `CharacterSet.uppercaseLetters` and `CharacterSet.lowercaseLetters` (Unicode General Categories Lu and Lt).
+        /// The conversion to lower case uses `Locale.system`, also known as the ICU "root" locale. This means the result is consistent regardless of the current user's locale and language preferences.
+        ///
+        /// Converting from camel case to snake case:
+        /// 1. Splits words at the boundary of lower-case to upper-case
+        /// 2. Inserts `_` between words
+        /// 3. Lowercases the entire string
+        /// 4. Preserves starting and ending `_`.
+        ///
+        /// For example, `oneTwoThree` becomes `one_two_three`. `_oneTwoThree_` becomes `_one_two_three_`.
+        ///
+        /// - Note: Using a key encoding strategy has a nominal performance cost, as each string key has to be converted.
+        case convertToSnakeCase
+        
+        /// Provide a custom conversion to the key in the encoded JSON from the keys specified by the encoded types.
+        /// The full path to the current encoding position is provided for context (in case you need to locate this key within the payload). The returned key is used in place of the last component in the coding path before encoding.
+        case custom(([CodingKey]) -> CodingKey)
+    }
+    
+    
+    /// The strategy to use for encoding keys. Defaults to `.useDefaultKeys`.
+    open var keyEncodingStrategy: KeyEncodingStrategy = .useDefaultKeys
+}
+```
+
+## Detailed design
+
+The strategy enum allows developers to pick from common actions of converting to and from `snake_case` to the Swift-standard `camelCase`. The implementation is intentionally simple, because we want to make the rules predictable.
+
+Converting from snake case to camel case:
+
+1. Capitalizes the word starting after each `_`
+2. Removes all `_`
+3. Preserves starting and ending `_` (as these are often used to indicate private variables or other metadata).
+
+For example, `one_two_three` becomes `oneTwoThree`. `_one_two_three_` becomes `_oneTwoThree_`.
+
+Converting from camel case to snake case:
+
+1. Splits words at the boundary of lower-case to upper-case
```

Comment:
If I undestand correctly, this implies `imageURL` becomes `image_URL`, right? Or rather `image_url` I guess.
Can we add examples with acronyms to make it clear?

Reply:
ah, right
>Lowercases the entire string
so `image_url` would be the one, yes?

Reply:
Yah, check out the test cases at https://github.com/apple/swift/pull/12779/ for some additional examples.